### PR TITLE
Do not try to infer the logprob of broadcasted or advanced indexed variables

### PR DESCRIPTION
**Pull request by ricardoV94**
This PR removes an invalid rewrite from the logprob submodule

It also suppresses a NotImplementedError in the Mixture rewrite. Individual rewrites should not judge whether logprob for a graph is implemented or not, because other rewrites might be able to handle it.

### CODE REVIEW

File: pymc/tests/logprob/test_transforms.py [hunk 1]
```diff
@@ -709,13 +709,13 @@ def test_discrete_rv_multinary_transform_fails():
         joint_logprob({y_rv: y_rv.clone()})
 
 
-@pytest.mark.xfail(reason="Check not implemented yet, see #51")
+@pytest.mark.xfail(reason="Check not implemented yet")
```
Comment (ricardoV94) [inline pymc/tests/logprob/test_transforms.py:781]: Add companion test with explicit `broadcast_to` which now fails after this PR

File: pymc/logprob/mixture.py [hunk 2]
```diff
@@ -407,6 +418,8 @@ def logprob_MixtureRV(
 
     assert len(indices) > 0
 
+    # FIXME: This logic does not account for repeated indexes, in which case the Mixture values are not independent.
+    #  It over-counts the logp terms for values that were repeated during indexing
     if len(indices) > 1 or indices[0].ndim > 0:
```
Comment (ricardoV94) [inline pymc/logprob/mixture.py]: Make sure we have tests for still supported multiple indexes

```
Normal.dist(size=(2, 3)[0, 0]
```

Comment (ricardoV94) [inline pymc/logprob/mixture.py]: Remove or branch (Advanced Indexing)?

**Comment by zaxtax** [comment]
Are we guaranteed to always be able to later rewrite this safely into something that is a proper logprob?

**Comment by ricardoV94** [comment]
> Are we guaranteed to always be able to later rewrite this safely into something that is a proper logprob?

There are two types of logps that are relevant here. An RV that is broadcasted directly and an RV where we are doing advanced indexing (i.e. mixture)

For the former, we would need to perform symbolic unbroadcasting of the value variable, and then evaluate the base RV at that value

```python
import pytensor.tensor as pt

def logp_broadcast_to(value, base_rv, broadcast_shape):
  return logp(base_rv, unbroadcast(value, broadcast_shape))
```

I am not sure what the `unbroadcast` function would look like, but it should do the following (in pseudo-code), when valid, and raise otherwise.
```python
np.testing.assert_array_equal(unbroadcast(broadcast_to(value, shape), shape), value)
```

For the second, we would need to somehow only evaluate the logp at the unique indexes. Something like 

```python
def logp_mixture_advanced_indexing(value, base_rvs, indexes):
  unique_indexes = get_unique_indexes(unique_indexes)
  # And either ignore or assert the value shoud be the same for repeated indexes
  return raveled_mixture_rv(value[unique_indexes], base_rvs, unique_indexes)
```

Neither seems trivial to implement, which is why this PR does not attempt to do so.

**Comment by zaxtax** [comment]
Just to clarify, are there rewrites this allows that weren't allowed before? It seems to be upon closer reading more restrictive about where rewrites are allowed to happen. In either case what we do for Case 1 vs. Case 2 should probably be different.

File: pymc/logprob/mixture.py [hunk 3]
```diff
@@ -300,13 +300,27 @@ def mixture_replace(fgraph, node):
 
     old_mixture_rv = node.default_output()
 
-    mixture_res, join_axis = get_stack_mixture_vars(node)
+    try:
+        mixture_res, join_axis = get_stack_mixture_vars(node)
+    except NotImplementedError:
```
Comment (zaxtax) [inline pymc/logprob/mixture.py]: This seems like a code smell. Would it be unreasonable to refactor `get_stack_mixture_vars` to not raise this exception?

File: pymc/logprob/mixture.py [hunk 4]
```diff
@@ -404,6 +418,8 @@ def logprob_MixtureRV(
 
     assert len(indices) > 0
 
+    # FIXME: This logic does not account for repeated indexes, in which case the Mixture values are not independent.
+    #  It over-counts the logp terms for values that were repeated during indexing
```
Comment (zaxtax) [inline pymc/logprob/mixture.py]: Should this have a unit test when it's resolved?

File: pymc/logprob/mixture.py (hunk 3 above)
Comment (ricardoV94) [reply pymc/logprob/mixture.py]: I refactored it to avoid raising the exception

File: pymc/logprob/mixture.py (hunk 4 above)
Comment (ricardoV94) [reply pymc/logprob/mixture.py]: I don't know if we will ever resolve it. For now we don't allow creating such MixtureRVs in the first place. The ones that hit this branch don't do AdvancedIndexing

**Comment by zaxtax** [comment]
> > Just to clarify, are there rewrites this allows that weren't allowed before? It seems to be upon closer reading more restrictive about where rewrites are allowed to happen. In either case what we do for Case 1 vs. Case 2 should probably be different.
> 
> Yes this PR restricts the type of rewrites that are possible, and does not enable any new ones. I am not sure we need to bother with Case1, but we can explore it. Case2 should definitely be allowed (at least the case that does not have repeated indexes). Would you mind opening an issue to track that?

I don't think I understand it well-enough to write a good issue for it.

**Comment by ricardoV94** [comment]
Converted to a draft, as this and #6397 touch on similar issues

**Comment by zaxtax** [comment]
LGTM!

**Pull request merged**